core/src/main/java/hudson/util/AtomicFileWriter.java

```diff
@@ -58,27 +64,51 @@ public AtomicFileWriter(File f) throws IOException {
     }
 
     /**
-     * @param encoding
-     *      File encoding to write. If null, platform default encoding is chosen.
+     * @param encoding File encoding to write. If null, platform default encoding is chosen.
+     *
+     * @deprecated Use {@link #AtomicFileWriter(Path, Charset)}
      */
-    public AtomicFileWriter(File f, String encoding) throws IOException {
-        File dir = f.getParentFile();
-        try {
-            dir.mkdirs();
-            tmpFile = File.createTempFile("atomic",null, dir);
-        } catch (IOException e) {
-            throw new IOException("Failed to create a temporary file in "+ dir,e);
-        }
-        destFile = f;
-        if (encoding==null)
-            encoding = Charset.defaultCharset().name();
+    @Deprecated
+    public AtomicFileWriter(@Nonnull File f, @Nullable String encoding) throws IOException {
+        this(toPath(f), encoding == null ? Charset.defaultCharset() : Charset.forName(encoding));
+    }
+
+    /**
+     * Wraps potential {@link java.nio.file.InvalidPathException} thrown by {@link File#toPath()} in an
+     * {@link IOException} for backward compatibility.
+     *
+     * @param file
+     * @return the path for that file
+     * @see File#toPath()
+     */
+    private static Path toPath(@Nonnull File file) throws IOException {
         try {
-            core = new BufferedWriter(new OutputStreamWriter(Files.newOutputStream(tmpFile.toPath()), encoding));
+            return file.toPath();
         } catch (InvalidPathException e) {
             throw new IOException(e);
         }
     }
 
+    /**
+     * @param destinationPath the destination path where to write the content when committed.
+     * @param charset File charset to write.
+     */
+    public AtomicFileWriter(@Nonnull Path destinationPath, @Nonnull Charset charset) throws IOException {
+        if (charset == null) { // be extra-defensive if people don't care
+            throw new IllegalArgumentException("charset is null");
+        }
+        this.destPath = destinationPath;
+        Path dir = this.destPath.getParent();
+        try {
+            Files.createDirectories(dir);
+            tmpPath = Files.createTempFile(dir, "atomic", "tmp");
```

Comment:
@batmat what's the lifecycle of this? Be weary that certain Redhat distributions periodically destroy files in the tmp directory [1].

[1]

The file /etc/cron.daily/tmpwatch defines the way tmpwatch is called.

```
#! /bin/sh
flags=-umc
/usr/sbin/tmpwatch "$flags" -x /tmp/.X11-unix -x /tmp/.XIM-unix \
    -x /tmp/.font-unix -x /tmp/.ICE-unix -x /tmp/.Test-unix \
    -X '/tmp/hsperfdata_*' 10d /tmp
```

Reply:
If it is short lived (for the operation) it is fine, just thought I would share the above experience :)

Reply:
This is not in the temp directory but the same directory as the target file (generally somewhere under JENKINS_HOME)

Reply:
Even better!